Sources/Testing/Attachments/Test.Attachable.swift

```diff
@@ -0,0 +1,171 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+@_spi(Experimental)
+extension Test {
+  /// A protocol describing a type that can be attached to a test report or
+  /// written to disk when a test is run.
+  ///
+  /// To attach an attachable value to a test report or test run output, use it
+  /// to initialize a new instance of ``Test/Attachment``, then call
+  /// ``Test/Attachment/attach()``. An attachment can only be attached once.
+  ///
+  /// The testing library provides default conformances to this protocol for a
+  /// variety of standard library types. Most user-defined types do not need to
+  /// conform to this protocol.
+  ///
+  /// A type should conform to this protocol if it can be represented as a
+  /// sequence of bytes that would be diagnostically useful if a test fails.
+  public protocol Attachable: ~Copyable {
+    /// An estimate of the number of bytes of memory needed to store this value
+    /// as an attachment.
+    ///
+    /// The testing library uses this property to determine if an attachment
+    /// should be held in memory or should be immediately persisted to storage.
+    /// Larger attachments are more likely to be persisted, but the algorithm
+    /// the testing library uses is an implementation detail and is subject to
+    /// change.
+    ///
+    /// The value of this property is approximately equal to the number of bytes
+    /// that will actually be needed, or `nil` if the value cannot be computed
+    /// efficiently. The default implementation of this property returns `nil`.
+    ///
+    /// - Complexity: O(1) unless `Self` conforms to `Collection`, in which case
+    ///   up to O(_n_) where _n_ is the length of the collection.
+    var estimatedAttachmentByteCount: Int? { get }
```

Comment:
```suggestion
    var estimatedByteCount: Int? { get }
```

(This name can be debated more in subsequent PRs, just caught my eye on the second pass…)

Reply:
Needs `attachment` or similar to avoid naming conflicts/confusion.